Document Explicit/Implicit operators for IntPtr, UIntPtr, BigInteger

tompratt-AQ opened the pull request:
# Title
Document all Explicit and Implicit operators for IntPtr, UIntPtr, and BigInteger.

## Summary
Pulled information from MSDN to document all Explicit and Implicit operators in:
- System.IntPtr  (6 Explicit)
- System.UIntPtr  (6 Explicit)
- System.Numerics.BigInteger  (14 Explicit and 8 Implicit)

## Suggested Reviewers
@rpetrusha 

**Question for reviewer**
A few of these methods had the following message as part of the description on MSDN:
"This API is not CLS-compliant. "

Does this message need to be added to our description?  Or is it generated?

tompratt-AQ committed:
Document all Explicit and Implicit operators for IntPtr, UIntPtr, and BigInteger.

tompratt-AQ committed:
Fix code sample paths.

tompratt-AQ committed:
Fixed code sample path.

rpetrusha commented on xml/System.Numerics/BigInteger.xml:
```diff
@@ -2861,10 +2942,31 @@ value Mod 2 = 0
+## Remarks  
+ The overloads of the <xref:System.Numerics.BigInteger.op_Explicit(System.Decimal)~System.Numerics.BigInteger> method define the types to which or from which a <xref:System.Numerics.BigInteger> object can be converted. Language compilers do not perform this conversion automatically because it can involve data loss. Instead, they perform the conversion only if a casting operator (in C#) or a conve ...
+
+ Because the <xref:System.Numerics.BigInteger> value can be outside the range of the <xref:System.Double> data type, this operation is a narrowing conversion. If the conversion is unsuccessful, it does not throw an <xref:System.OverflowException>. Instead, if the <xref:System.Numerics.BigInteger> value is less than <xref:System.Double.MinValue>, the resulting <xref:System.Double> value is <xref:S ...
```
Two issues here:
- The xrefs to Double.Minvalue, Double.MaxValue (which isn't included yet -- see the next point), Double.PositiveInfinity, and Double.NegativeInfinity should all be `?displayProperty=fullName`.
This should be "greater than `<xref:System.Double.MaxValue?displayProperty=fullName>` instead of `<xref:System.Numerics.BigInteger>`.

rpetrusha commented on xml/System.Numerics/BigInteger.xml:
```diff
@@ -2728,10 +2728,26 @@ value Mod 2 = 0
+## Remarks  
+ Any fractional part of the `value` parameter is truncated before conversion.
+
+ The overloads of the Explicit(Decimal to BigInteger) method define the types to which or from which a <xref:System.Numerics.BigInteger> object can be converted. Because the conversion from <xref:System.Decimal> to <xref:System.Numerics.BigInteger> can involve truncating any fractional part of `value`, language compilers do not perform this conversion automatically. Instead, they perform the conv ...
```
Explicit(Decimal to BigInteger) should be `<xref:System.Numbers.BigInteger.op_Explicit%2A>`. That's probably true of this paragraph in each version of the overloads.

rpetrusha commented on xml/System.Numerics/BigInteger.xml:
```diff
@@ -3036,10 +3218,31 @@ value Mod 2 = 0
+## Remarks  
+ The overloads of the <xref:System.Numerics.BigInteger.op_Explicit(System.Decimal)~System.Numerics.BigInteger> method define the types to which or from which a <xref:System.Numerics.BigInteger> object can be converted. Language compilers do not perform this conversion automatically because it can involve data loss or a loss of precision. Instead, they perform the conversion only if a casting oper ...
+
+ Because the <xref:System.Numerics.BigInteger> value can be outside the range of the <xref:System.Single> data type, this operation is a narrowing conversion. If the conversion is unsuccessful, it does not throw an <xref:System.OverflowException>. Instead, if the <xref:System.Numerics.BigInteger> value is less than <xref:System.Single.MinValue>, the resulting <xref:System.Single> value is <xref:S ...
```
The xrefs to Single.MinValue, Single.MaxValue, Single.PositiveInfinity, and Single.NegativeInfinity should all be `?displayProperty=fullName`.

rpetrusha commented on xml/System.Numerics/BigInteger.xml:
```diff
@@ -2861,10 +2942,31 @@ value Mod 2 = 0
+
+ Because the <xref:System.Numerics.BigInteger> value can be outside the range of the <xref:System.Double> data type, this operation is a narrowing conversion. If the conversion is unsuccessful, it does not throw an <xref:System.OverflowException>. Instead, if the <xref:System.Numerics.BigInteger> value is less than <xref:System.Double.MinValue>, the resulting <xref:System.Double> value is <xref:S ...
+ 
+ The conversion of a <xref:System.Numerics.BigInteger> to a <xref:System.Double> may involve a loss of precision. In some cases, the loss of precision may cause the casting or conversion operation to succeed even if the <xref:System.Numerics.BigInteger> value is outside the range of the <xref:System.Double> data type. The following example provides an illustration. It assigns the maximum value of ...
```
Double.MaxValue should be `?displayProperty=fullName`

rpetrusha commented on xml/System.Numerics/BigInteger.xml:
```diff
@@ -3036,10 +3218,31 @@ value Mod 2 = 0
+
+ Because the <xref:System.Numerics.BigInteger> value can be outside the range of the <xref:System.Single> data type, this operation is a narrowing conversion. If the conversion is unsuccessful, it does not throw an <xref:System.OverflowException>. Instead, if the <xref:System.Numerics.BigInteger> value is less than <xref:System.Single.MinValue>, the resulting <xref:System.Single> value is <xref:S ...
+
+ The conversion of a <xref:System.Numerics.BigInteger> to a <xref:System.Single> may involve a loss of precision. In some cases, the loss of precision may cause the casting or conversion operation to succeed even if the <xref:System.Numerics.BigInteger> value is outside the range of the <xref:System.Single> data type. The following example provides an illustration. It assigns the maximum value of ...
```
Single.MaxValue should be `?displayProperty=fullName`

rpetrusha reviewed (changes requested):
@tompratt-AQ, the examples didn't build. See my comments.

tompratt-AQ committed:
Make changes per review.

tompratt-AQ commented:
@rpetrusha I made changes per your review comments for the editing issues. For the code examples, I checked that the paths and snippet tags are correct, and made sure there were 3 spaces at the end of the lines. I left the ~ in the paths, because other existing snippets in the topic that are appearing correctly use that character. Also, could you answer my question that I put under "Suggested Reviewers" up above? Need to know whether to include a line. Thanks!

tompratt-AQ committed:
Fixed bad links per build Warnings.

tompratt-AQ commented:
Fixed bad links for 2 of the topics that were listed in the Build warnings.

rpetrusha commented:
Sorry, @tompratt-AQ -- I missed your question. The boilerplate "This API is not CLS-compliant." was added automatically by the build, but no longer is on OPS. So yes, please add it where appropriate.

rpetrusha commented:
Sorry, @tompratt-AQ, there's one detail I forgot to mention. On MSDN, the text sometimes reads something like `equivalentCodeEntityM:System.Decimal.ToUInt64(System.Decimal)`. In that case, the text that you add should include the alternative API. So in this case, `This API is not CLS-compliant. The compliant alternative is <xref:System.Decimal.ToUInt64%2A>.`

--

tompratt-AQ commented:
@rpetrusha I sent you an email with a couple questions about adding the CLS compliant message. Thanks.

tompratt-AQ committed:
Added non CLS-compliant and alternative API messages.

tompratt-AQ commented:
@rpetrusha I added the non-compliance line with the compliant method to the Summary, and added the alternative method line that you suggested in the Remarks.

tompratt-AQ committed:
Added alternative method remarks for op_implicit.

tompratt-AQ commented:
@rpetrusha Please review the alternative method remarks and compliant method lines in the summaries for op_Explicit and op_Implicit. Thanks!

rpetrusha commented:
There are a number of broken links, @tompratt-AQ. So `The overloads of the <xref:System.Numerics.BigInteger.op_Explicit%28System.Decimal%29?displayProperty=fullName> method ` --> `The overloads of the <xref:System.Numerics.BigInteger.op_Explicit%2A?displayProperty=fullName> method` . And the broken links to constructor overloads are broken because `displayProperty` should be preceded by `?`.  Finally, could you change "For languages that do not support explicit operators..." to "For language that do not support custom operators..." Otherwise, this looks good.

tompratt-AQ committed:
Addressed broken links and suggested changes.

tompratt-AQ commented:
@rpetrusha Thanks for catching those, Ron. I made your suggested change, and I don't see any new broken links on the report, so this *might* be ready.

rpetrusha commented:
This looks good. Thanks, @tompratt-AQ.
